mitmproxy/addons/share.py

```diff
@@ -0,0 +1,73 @@
+import typing
+import random
+import datetime
+import time
+import _io
+import http.client
+
+from mitmproxy import command
+from mitmproxy import io
+from mitmproxy import ctx
+from mitmproxy import flow
+
+
+class Share:
+    def encode_multipart_formdata(self, filename, content):
+        params = {"key": filename, "acl": "bucket-owner-full-control", "Content-Type": "application/octet-stream"}
+        LIMIT = b'---------------------------198495659117975628761412556003'
+        CRLF = b'\r\n'
+        l = []
+        for (key, value) in params.items():
+            l.append(b'--' + LIMIT)
+            l.append(b'Content-Disposition: form-data; name="%b"' % key.encode("utf-8"))
+            l.append(b'')
+            l.append(value.encode("utf-8"))
+        l.append(b'--' + LIMIT)
+        l.append(b'Content-Disposition: form-data; name="file"; filename="%b"' % filename.encode("utf-8"))
+        l.append(b'Content-Type: application/octet-stream')
+        l.append(b'')
+        l.append(content)
+        l.append(b'--' + LIMIT + b'--')
+        l.append(b'')
+        body = CRLF.join(l)
+        content_type = b'multipart/form-data; boundary=%b' % LIMIT
+        return content_type, body
+
+    def post_multipart(self, host, filename, content):
+        content_type, body = self.encode_multipart_formdata(filename, content)
+        conn = http.client.HTTPConnection(host, 80)
+        headers = {'content-type': content_type, 'content-length': str(len(body))}
+        try:
+            conn.request("POST", "", body, headers)
+        except http.client.CannotSendRequest:
+            return 'We failed to reach a server.'
+        try:
+            conn.getresponse()
+        except http.client.RemoteDisconnected:
+            return 'The server couldn\'t fulfill the request.'
+        else:
+            conn.close()
```

Comment:
The logic here is flawed - we should always call close, no matter if the connection succeeded. 

The number of comments on this function is getting a bit overboard. In total it should look roughly like this:
```python
    def post_multipart(self, host: str, filename: str, content: bytes) -> str:
        """
        Upload flows to the specified S3 server.

        Returns:
            - The share URL, if upload is successful.
        Raises:
            - IOError, otherwise.
        """
        content_type, body = self.encode_multipart_formdata(filename, content)
        headers = {'content-type': content_type}

        conn = http.client.HTTPSConnection(host)

        try:
            conn.request("POST", "/", body, headers)
            resp = conn.getresponse()
        finally:
            conn.close()

        if resp.status != 200:
            if resp.reason:
                reason = resp.reason
            else:
                 reason = status_codes.RESPONSES.get(resp.status, str(resp.status))
            raise IOError(reason)
        return "https://share.mitmproxy.org/%s" % filename
```

`share` needs to properly catch IOErrors now of course.

Reply:
Hello sir, I couldn't understand why are we checking if `resp.reason` is empty or not, don't we always receive a reason?

Reply:
Not always, e.g. HTTP/2.0 does not include a reason in responses. It would be awkward to have an error message that reads like this: "Error uploading flow: "